refactor: more robust tagging index sync as sender

benesjan opened the pull request:
In this PR I implement the approach that Nico originally described in [this post](https://forum.aztec.network/t/on-note-discovery-and-index-coordination/7165) for sender tagging index sync (for recipient it will be done in a followup PR).

To summarize I do the following:
1. Associate the tagging index with a hash of a tx in which the corresponding tax was used,
2. I track the status of the given tx in the tagging data provider,
3. tagging index status can be either PENDING, FINALIZED or DROPPED_OR_REVERTED (this status is only symbolic as when we realize a corresponding tx has been dropped or reverted we just delete the pending index),
4. when choosing the next index to use when sending a private log I choose `HIGHEST_PENDING_INDEX + 1` (or `HIGHEST_FINALIZED_INDEX + 1` if there is no pending index),
5. if the chosen index is further than WINDOW_LENGTH away from the HIGHEST_FINALIZED_INDEX an error is thrown. This should be a good guarantee of us never losing a log because we are always looking for a WINDOW_LENGTH of logs.

# Notes for reviewer
- I separated `TaggingDataProvider` into `SenderTaggingDataProvider` and `RecipientTaggingDataProvider` because the algorithms are completely disjoint and it makes the code clearer.
- Moved the `syncTaggedLogsAsSender` function to `pxe/src/tagging` directory and renamed it as `syncSenderTaggingIndexes`. Now `PrivateExecutionOracle` just calls this function. This is a step in the direction of deprecating `PXEOracleInterface` and it allowed me to write a reasonably clear unit tests.

# Issues created in this PR
- https://github.com/AztecProtocol/aztec-packages/issues/17775
- https://github.com/AztecProtocol/aztec-packages/issues/17776

benesjan commented:
* **#19290** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19290?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19289** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19289?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19233** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19233?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19125** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19125?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19161** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19161?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19158** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19158?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19157** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19157?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19030** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19030?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17611** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17611?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17611?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#17568** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17568?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17534** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17534?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17445** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17445?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17451** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17451?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on yarn-project/end-to-end/src/e2e_pending_note_hashes_contract.test.ts:
```diff
@@ -284,6 +284,8 @@ describe('e2e_pending_note_hashes_contract', () => {
 
   it('Should handle overflowing the kernel data structures in nested calls', async () => {
+    // This test verifies that a transaction can emit more notes than MAX_NOTE_HASHES_PER_TX without failing, since
+    // the notes are nullified and will be squashed by the kernel reset circuit.
```
Unrelated comment I added when debugging this test.

benesjan commented on yarn-project/pxe/src/contract_function_simulator/execution_data_provider.ts:
```diff
@@ -227,21 +227,13 @@ export interface ExecutionDataProvider {
-   * first field of the private log which corresponds to the tag).
-   */
+  // docs available in sync_sender_tagging_indexes.ts
   syncTaggedLogsAsSender(secret: DirectionalAppTaggingSecret, contractAddress: AztecAddress): Promise<void>;
```
Felt unnecessary to have the docs copy here especially when PXEOracleInterface will be dropped.

benesjan commented on yarn-project/pxe/src/contract_function_simulator/pxe_oracle_interface.test.ts:
```diff
@@ -219,88 +223,6 @@ describe('PXEOracleInterface', () => {
       expect(aztecNode.getLogsByTags.mock.calls.length).toBe(2);
     });
 
-    it('should sync tagged logs as senders', async () => {
```
This test was replaced with tests in `sync_sender_tagging_indexes.test.ts`.

benesjan commented on yarn-project/pxe/src/contract_function_simulator/pxe_oracle_interface.ts:
```diff
@@ -336,69 +338,14 @@ export class PXEOracleInterface implements ExecutionDataProvider {
+  // Note: I think syncSenderTaggingIndexes would be a better name as it better communicates the purpose of
+  // the function - we care about finding the indexes but we don't care about the log contents.
+  // TODO(benesjan): Rename the function to syncSenderTaggingIndexes if reviewer agrees.
   public async syncTaggedLogsAsSender(
```
The whole algorithm has been replaced so it doesn't make sense to look at the deleted code. The new implementation is in the `syncSenderTaggingIndexes` function. Does the reviewer agree with the new name? I think the new name is better because the whole point of this is not to sync the logs (in a sense that we don't care about the log contents) but to sync the tagging indexes.

Anyway I think it would make sense to not rename the function here but to instead just go ahead and drop the whole PXEOracleInterface class along with the ExecutionDataProvider interface.

benesjan commented on yarn-project/pxe/src/contract_function_simulator/pxe_oracle_interface.ts:
```diff
@@ -336,80 +338,19 @@ export class PXEOracleInterface implements ExecutionDataProvider {
+  // TODO(benesjan): Replace this implementation of this function with one implementing an approach similar
+  // to syncSenderTaggingIndexes. Not done yet due to re-prioritization to devex and this doesn't directly affect
+  // devex.
   public async syncTaggedLogs(
```
The docs here were a duplicate from ExecutionDataProvider so I just dropped and left here this TODO explaining why the approaches are different.

benesjan commented on yarn-project/pxe/src/storage/tagging_data_provider/recipient_tagging_data_provider.ts:
```diff
@@ -0,0 +1,86 @@
+import { toArray } from '@aztec/foundation/iterable';
```
This file is basically just a copy-paste of the recipient-sync related functionality from the original TaggingDataProvider so I don't think you really need to review this. It will also get basically replaced once the recipient sync is being refactored as well.

benesjan commented on yarn-project/pxe/src/storage/tagging_data_provider/sender_tagging_data_provider.test.ts:
```diff
@@ -0,0 +1,444 @@
+
+import { SenderTaggingDataProvider } from './sender_tagging_data_provider.js';
+
+describe('SenderTaggingDataProvider', () => {
```
The unit tests didn't exist before for the tagging data provider so I implemented them here. For the recipient tagging data provider I will implement them once the recipient sync algo is reworked.

benesjan commented on yarn-project/pxe/src/storage/tagging_data_provider/sender_tagging_data_provider.ts:
```diff
@@ -0,0 +1,210 @@
+import { toArray } from '@aztec/foundation/iterable';
```
None of the functionality from the original tagging data provider has been copied here as now the sync algo works completely differently.

benesjan commented on yarn-project/pxe/src/block_synchronizer/block_synchronizer.ts:
```diff
@@ -65,7 +65,9 @@ export class Synchronizer implements L2BlockStreamEventHandler {
-        await this.taggingDataProvider.resetNoteSyncData();
+        // Note: This is now unnecessary for the sender tagging data provider because the new algorithm handles reorgs.
+        // TODO(benesjan): Drop this on recipient tagging algo refactor.
+        await this.recipientTaggingDataProvider.resetNoteSyncData();
```
This is a big advantage of the new approach as now reorgs are not a problem as the algorithm will automatically handle it by dropping the pending indexes of the dropped txs

benesjan commented on yarn-project/pxe/src/tagging/sync/sync_sender_tagging_indexes.ts:
```diff
@@ -0,0 +1,158 @@
+// tagging secret. If we get more tag indexes consumed than this window, an error is thrown in `PXE::proveTx` function.
+// This is set to a larger value than MAX_PRIVATE_LOGS_PER_TX (currently 32) because there could be more than
+// MAX_PRIVATE_LOGS_PER_TX indexes consumed in case the logs are squashed. This happens when the log contains a note
+// and the note is nullified in the same tx.
```
This is quite unfortunate but there is no way around it because we cannot squash the used indexes as that would require knowing which logs will end up being squashed ahead of them being actually emitted.

benesjan commented on yarn-project/pxe/src/pxe.ts:
```diff
@@ -478,50 +484,52 @@ export class PXE {
+   * After registering a new sender, the PXE will sync private logs that are tagged with this sender's address.
+   * Will do nothing if the address is already registered.
    *
-   * @param address - Address of the user to add to the address book
-   * @returns The address address of the account.
+   * @param sender - Address of the sender to register.
+   * @returns The address of the sender.
+   * TODO: It's strange that we return the address here and I (benesjan) think we should drop the return value.
    */
-  public async registerSender(address: AztecAddress): Promise<AztecAddress> {
+  public async registerSender(sender: AztecAddress): Promise<AztecAddress> {
     const accounts = await this.keyStore.getAccounts();
-    if (accounts.includes(address)) {
-      this.log.info(`Sender:\n "${address.toString()}"\n already registered.`);
-      return address;
+    if (accounts.includes(sender)) {
+      this.log.info(`Sender:\n "${sender.toString()}"\n already registered.`);
+      return sender;
     }
 
-    const wasAdded = await this.taggingDataProvider.addSenderAddress(address);
+    const wasAdded = await this.recipientTaggingDataProvider.addSenderAddress(sender);
 
     if (wasAdded) {
-      this.log.info(`Added sender:\n ${address.toString()}`);
+      this.log.info(`Added sender:\n ${sender.toString()}`);
     } else {
-      this.log.info(`Sender:\n "${address.toString()}"\n already registered.`);
+      this.log.info(`Sender:\n "${sender.toString()}"\n already registered.`);
     }
 
-    return address;
+    return sender;
```
There are no important changes here but I just didn't like the docs and naming here much so I fixed it as it's somewhat related to this PR. Also the docs were a bit stale as it just talked about receiving notes but private logs can also contain other kind of messages (like events).

benesjan commented on yarn-project/pxe/src/pxe.ts:
```diff
@@ -478,50 +484,52 @@ export class PXE {
     }
 
-    return address;
+    return sender;
   }
 
   /**
-   * Retrieves the addresses stored as senders on this PXE.
-   * @returns An array of the senders on this PXE.
+   * Retrieves senders registered in this PXE.
+   * @returns Senders registered in this PXE.
```
Didn't like that we describe the return type here when it's clearly visible from the function return types.

benesjan commented on yarn-project/pxe/src/pxe.ts:
```diff
@@ -478,50 +484,52 @@ export class PXE {
+    return this.recipientTaggingDataProvider.getSenderAddresses();
   }
 
   /**
-   * Removes a sender in the address book.
+   * Removes a sender registered in this PXE.
+   * @param sender - The address of the sender to remove.
    */
-  public async removeSender(address: AztecAddress): Promise<void> {
-    const wasRemoved = await this.taggingDataProvider.removeSenderAddress(address);
+  public async removeSender(sender: AztecAddress): Promise<void> {
+    const wasRemoved = await this.recipientTaggingDataProvider.removeSenderAddress(sender);
 
     if (wasRemoved) {
-      this.log.info(`Removed sender:\n ${address.toString()}`);
+      this.log.info(`Removed sender:\n ${sender.toString()}`);
     } else {
-      this.log.info(`Sender:\n "${address.toString()}"\n not in address book.`);
+      this.log.info(`Sender:\n "${sender.toString()}"\n not registered in PXE.`);
```
Here it talked about an address book and that's probably because in the RecipientTaggingData provider we call that map an address book. Mentioning it here seems unnecessarily weird.

benesjan commented on yarn-project/pxe/src/pxe.ts:
```diff
@@ -743,14 +751,33 @@ export class PXE {
+          }
+
+          // TODO(benesjan): The following is an expensive operation. Figure out a way to avoid it.
+          const txHash = (await txProvingResult.toTx()).txHash;
```
@Thunkar we were discussing merging proveTx with sendTx. Am I correct in assuming that if that gets done this will not be a problem as we'll be able to move this to the end of the newly introduced PXE::sendTx function where we'll have access to the "real" final `Tx` object?

benesjan commented on yarn-project/txe/src/state_machine/archiver.ts:
```diff
@@ -107,8 +107,21 @@ export class TXEArchiver extends ArchiverStoreHelper implements L2BlockSource {
 
-  public getL2Tips(): Promise<L2Tips> {
-    throw new Error('TXE Archiver does not implement "getL2Tips"');
+  public async getL2Tips(): Promise<L2Tips> {
```
I needed to implement this because it is called by the `syncSenderTaggingIndexes` function which is expected to be triggered even in the TXE flows because the `privateGetNextAppTagAsSender` oracle is called whenever we emit a log.

benesjan commented on yarn-project/txe/src/state_machine/dummy_p2p_client.ts:
```diff
@@ -74,7 +74,9 @@ export class DummyP2P implements P2P {
   }
 
   public getTxStatus(_txHash: TxHash): Promise<'pending' | 'mined' | undefined> {
```
Same as `getL2Tips` - also called by `syncSenderTaggingIndexes`.

benesjan commented on yarn-project/pxe/src/contract_function_simulator/oracle/private_execution.test.ts:
```diff
@@ -107,6 +109,8 @@ describe('Private Execution test suite', () => {
 
   let executionDataProvider: MockProxy<ExecutionDataProvider>;
+  let senderTaggingDataProvider: MockProxy<SenderTaggingDataProvider>;
+  let aztecNode: MockProxy<AztecNode>;
```
The `getLastUsedIndexAsSender` and `syncTaggedLogsAsSender` have been dropped from PXEOracleInterface so I needed to update the mocks here as the node and tagging data provider are accessed directly by the `PrivateExecutionOracle`.

benesjan commented on yarn-project/pxe/src/contract_function_simulator/pxe_oracle_interface.ts:
```diff
@@ -336,80 +337,9 @@ export class PXEOracleInterface implements ExecutionDataProvider {
     }));
   }
 
-  public async syncTaggedLogsAsSender(
```
This function was dropped and re-implemented in `sync_sender_tagging_indexes.ts` file. The new function doesn't share any code with this.

nventuro commented on yarn-project/pxe/src/contract_function_simulator/oracle/private_execution_oracle.ts:
```diff
@@ -224,11 +225,58 @@ export class PrivateExecutionOracle extends UtilityExecutionOracle implements IP
+            `The node we are trying to sync indexes from is not aware of our anchor block with hash ${anchorBlockHash}. This should never happen if the sync is working correctly. Please send a bug report to the Aztec team.`,
+          );
+        }
+      }
```
As discussed, the `anchorBlock` should instead be a param in most (all?) node queries.

benesjan replied:
Reverted this and agree that having it be part of the queries is a much better idea 👍

nventuro commented on yarn-project/pxe/src/storage/tagging_data_provider/sender_tagging_data_provider.test.ts:
```diff
@@ -0,0 +1,444 @@
+      const preTag: PreTag = { secret: secret1, index: 5 };
+
+      await taggingDataProvider.storePendingIndexes([preTag], txHash);
+      await taggingDataProvider.storePendingIndexes([preTag], txHash);
```
What if we threw here?

benesjan replied:
That would be incorrect because this is expected to happen:

```
   * @remarks Ignores the index if the same preTag + txHash combination already exists in the db with the same index.
   * This is expected to happen because whenever we start sync we start from the last finalized index and we can have
   * pending indexes already stored from previous syncs.
```

nventuro commented on yarn-project/pxe/src/storage/tagging_data_provider/sender_tagging_data_provider.ts:
```diff
@@ -0,0 +1,210 @@
+import { TxHash } from '@aztec/stdlib/tx';
+
+/**
+ * Data provider of tagging data used when syncing the sender tagging indexes. The recipient alternative of this class
```
wdym alternative? Did you mean to say something like 'counterpart'?

benesjan replied:
me no speaks

https://github.com/AztecProtocol/aztec-packages/pull/17611/commits/bb588b359ad051018f821bffcf0925f881ded0fd

nventuro commented on yarn-project/pxe/src/storage/tagging_data_provider/sender_tagging_data_provider.ts:
```diff
@@ -0,0 +1,210 @@
+/**
+ * Data provider of tagging data used when syncing the sender tagging indexes. The recipient alternative of this class
+ * is called RecipientTaggingDataProvider. We have the providers separate for the sender and recipient because
+ * the algorithms are completely disjoint and there is not data reuse between the 2.
```
```suggestion
 * the algorithms are completely disjoint and there is not data reuse between the two.
```

nventuro commented on yarn-project/pxe/src/storage/tagging_data_provider/sender_tagging_data_provider.ts:
```diff
@@ -0,0 +1,210 @@
+  // contained the private logs with tags corresponding to these indexes has not been finalized yet. We don't store
+  // just the highest index because it could happen that some of the transactions are dropped and then we need
+  // the information about the lower pending indexes. But we store just one index per secret-txHash pair as only
+  // the highest index in the given tx is relevant for future index choice when sending a private log.
```
```suggestion
  // Stores the pending indexes for each directional app tagging secret. Pending here means that the tx that
  // contained the private logs with tags corresponding to these indexes has not been finalized yet.
  //
  // We don't store just the highest index because if their transaction is dropped we'd then need the
  // information about the lower pending indexes. For each secret-tx pair however we only store the largest index used in that tx, since the smaller ones are irrelevant due to tx atomicity (this assumes no
  // logs are used in the non-revertible phase).
```

>this assumes no logs are used in the non-revertible phase

I think this is true?

benesjan replied:
Yes, your changes improved it 👍 

Good point on the non-revertible phase logs. That seems like a problem we need to tackle. Would create an issue and link it here, ok?

benesjan replied:
Update: Just noticed I already have an [issue](https://github.com/AztecProtocol/aztec-packages/issues/17615) for dropping the non-revertible phase logs.

nventuro commented on yarn-project/pxe/src/storage/tagging_data_provider/sender_tagging_data_provider.ts:
```diff
@@ -0,0 +1,210 @@
+  /**
+   * Stores pending indexes.
+   * @remarks Ignores the index if the same preTag + txHash combination already exists in the db with the same index.
+   * This is expected to happen because whenever we start sync we start from the last finalized index and we can have
+   * pending indexes already stored from previous syncs.
```
I dont understand why we see the same one

benesjan replied:
Because whenever we initiate sync we start from the latest finalized index which means that it can happen that we load the same pending index multiple times.

And we start from latest finalized because pending indexes might have been dropped and we need to figure that out.

Do you think the comment is insufficient? This:
"because whenever we start sync we start from the last finalized index"

is quite clear, no?

nventuro replied:
Ah I see. Is it because we don't skip checking any indices for which we've already found a message?

benesjan replied:
Exactly. We stop checking for a given pending index only if it's finalized, some bigger pending index is finalized or the tx with the pending index is dropped.

nventuro commented on yarn-project/pxe/src/storage/tagging_data_provider/sender_tagging_data_provider.ts:
```diff
@@ -0,0 +1,210 @@
+   * pending indexes already stored from previous syncs.
+   * @param preTags - The pre-tags containing the directional app tagging secrets and the indexes that are to be
+   * stored in the db.
+   * @param txHash - The hash of the pending tx that used the given pre-tags to compute private log tags.
```
```suggestion
   * @param txHash - The tx in which the pretags were used in private logs.
```

nventuro commented on yarn-project/pxe/src/storage/tagging_data_provider/sender_tagging_data_provider.ts:
```diff
@@ -0,0 +1,210 @@
+   * @param preTags - The pre-tags containing the directional app tagging secrets and the indexes that are to be
+   * stored in the db.
+   * @param txHash - The hash of the pending tx that used the given pre-tags to compute private log tags.
+   * @throws If any two pre-tags contain the same directional app tagging secret. This is enforced because we care
+   * only about the highest index for a given secret that was used in the tx. Hence this check is a good way to catch
+   * bugs.
```
So the caller needs to make it so they only pass the largest index for a given secret? Why not do that here?

benesjan replied:
There are 2 paths in which this function is called:

**First path** - storing here the indexes once a tx is executed in this PXE

In this case we already need to track the highest index and we do so in `ExecutionTaggingIndexCache`.


**Second path** - storing here the indexes when performing sender tagging index sync

I checked if finding the highest index in the data provider instead of in the `loadAndStoreNewTaggingIndexes` could simplify it but I am came to the conclusion that not because in `loadAndStoreNewTaggingIndexes` we already need to perform transformation on the data and that is because there we get transaction hashes for a given pre tag and in the store here we need it inverted - we need pre tag for a given transaction.


Anyway, did a minor refactor of the related functionality in https://github.com/AztecProtocol/aztec-packages/pull/17611/commits/9276498bb86449968c9d4d0fda7a218b8ef48e60

nventuro commented on yarn-project/pxe/src/storage/tagging_data_provider/sender_tagging_data_provider.ts:
```diff
@@ -0,0 +1,210 @@
+
+    for (const { secret, index } of preTags) {
+      const secretStr = secret.toString();
+      const existing = (await this.#pendingIndexes.getAsync(secretStr)) ?? [];
```
Can we make these fns not take strings, so that we don't need to convert?

nventuro replied:
I'd move `existing` below since it's not used til later.

benesjan replied:
> Can we make these fns not take strings, so that we don't need to convert?

This is an issue with the underlying kv-store implementation that was done by alpha team. We need to use primitive types here because with reference type we would not get a key match. I think it would make sense to refactor this as I don't think this behavior is ever desirable. Not sure if there is some deeper issue making this refactor impractical though.

>I'd move existing below since it's not used til later.

https://github.com/AztecProtocol/aztec-packages/pull/17611/commits/a2fcb345b7300acc1e0584a7c8ffc3a8dc1e969f

nventuro replied:
Yeah, we'd need to do something like https://github.com/AztecProtocol/aztec-packages/pull/13702

nventuro commented on yarn-project/pxe/src/storage/tagging_data_provider/sender_tagging_data_provider.ts:
```diff
@@ -0,0 +1,210 @@
+              `a different index ${existingEntry.index} already exists for this secret-txHash pair`,
+          );
+        }
+        // If it exists with the same index, ignore the update (no-op)
```
What if it existed with a lower index? This seems to rely on the caller always passing the largest index, which seems maybe not great as that is greatly affecting the internal logic here. I'd strongly consider reworking that, or taking the highest index in here instead of outside, making the caller logic simpler.

benesjan replied:
Same response as [here](https://github.com/AztecProtocol/aztec-packages/pull/17611#discussion_r2620588207)

nventuro commented on yarn-project/pxe/src/storage/tagging_data_provider/sender_tagging_data_provider.ts:
```diff
@@ -0,0 +1,210 @@
+
+    // As the last used index we return the highest one from the pending indexes. Note that this might not technically
+    // be the last used index in case we sent logs from multiple PXEs in parallel but that's just a detail.
+    return Math.max(...pendingIndexes);
```
This assumes the pending indices will always be larger than the last finalized. This is wrong as transactions are not necessarily getting mined in the order in which we send them, e.g. index 1 is used in tx B with low fees, index 2 is used in tx B with high fees, B gets mined and eventually finalized, A sits on the mempool until tx fees drop.

benesjan replied:
When a transaction status is updated to finalized I prune any lower pending indexes for a given secret as they are no longer relevant.

I should have mentioned it in the comment here. Added it in https://github.com/AztecProtocol/aztec-packages/pull/17611/commits/f4f1665096d901134d8b0f69e58619d490549c51

There is a test case for this: 'prunes pending indexes with lower or equal index than finalized'

nventuro replied:
Not a fan of relying on it, but the comment is good.

nventuro commented on yarn-project/pxe/src/storage/tagging_data_provider/sender_tagging_data_provider.ts:
```diff
@@ -0,0 +1,210 @@
+   * @returns The last used index.
+   */
+  async getLastUsedIndex(secret: DirectionalAppTaggingSecret): Promise<number | undefined> {
+    const lastFinalizedIndex = await this.#lastFinalizedIndexes.getAsync(secret.toString());
```
Nitpicky, but creating variables that are unused until later makes the code unnecessarily harder to follow.

benesjan replied:
https://github.com/AztecProtocol/aztec-packages/pull/17611/commits/53b2037fa8fa72705d009cd3803c026f52bf5d9b

nventuro commented on yarn-project/pxe/src/storage/tagging_data_provider/sender_tagging_data_provider.ts:
```diff
@@ -0,0 +1,210 @@
+   * A tx that contained private logs with tags corresponding to some of the indexes returned from this data provider
+   * has been dropped so we delete the corresponding pending indexes. This results in the indexes being reused which
+   * results in the txs that reused the indexes to be linked with the dropped tx. We don't worry about that for now.
+   * @param txHash - The hash of the tx to drop the pending indexes for.
```
This is not a description of the function

benesjan replied:
https://github.com/AztecProtocol/aztec-packages/pull/17611/commits/8069b4a53a91ae814f2704a02407ecf0ae9bd8ab

nventuro commented on yarn-project/pxe/src/storage/tagging_data_provider/sender_tagging_data_provider.ts:
```diff
@@ -0,0 +1,210 @@
+    for (const secret of allSecrets) {
+      const pendingData = await this.#pendingIndexes.getAsync(secret);
+      if (pendingData) {
+        const filtered = pendingData.filter(item => item.txHash.toString() !== txHashStr);
```
Again there's unnecessary str conversion. You could not have `txHashStr` and just compare the tx hashes

benesjan replied:
As mentioned before `kv-store` needs primitive types when obtaining by key.

But the conversion on line 155 was indeed superfluous because `item.txHash` is already a string.

Cleaned that up in https://github.com/AztecProtocol/aztec-packages/pull/17611/commits/b64dbd3265d8ee5ec3923e1febfa896ed928860f

nventuro commented on yarn-project/pxe/src/storage/tagging_data_provider/sender_tagging_data_provider.ts:
```diff
@@ -0,0 +1,210 @@
+    const txHashStr = txHash.toString();
+    const allSecrets = await toArray(this.#pendingIndexes.keysAsync());
+
+    for (const secret of allSecrets) {
```
There's 4 levels of conditional nesting here, I'd try to find a simpler structure to express these ideas. e.g.

```
for secret of secrets {
  let pending = pending.get(secret)
  if !pending { cotninue }
  
  let matching = pending.filter(i.tx == tx).map(i.index);
  if matching.len == 0 { continue }
  else if matching.len > 1 {
    throw new error("this should never happen");
  }
  
  // updates
}
```

benesjan replied:
Good point.

Addressed in https://github.com/AztecProtocol/aztec-packages/pull/17611/commits/a7123fb84a56fdd1d4f4aafe893715b6951e4989

nventuro commented on yarn-project/pxe/src/storage/tagging_data_provider/sender_tagging_data_provider.ts:
```diff
@@ -0,0 +1,210 @@
+          .filter(item => item.txHash.toString() === txHashStr)
+          .map(item => item.index);
+        if (matchingIndexes.length === 1) {
+          let currentFinalized = await this.#lastFinalizedIndexes.getAsync(secret);
```
I had to read this repeatedly as i was getting confused by it. This whole section (`matching.len == 1`) is complicated enough to go into its own function. Some comments:
 - `currentFinalized` is a confusing name for something we also call `lastFinalized`.
 - the pruning is not well explained. In particular, it doesn't mention that even though all pending entries should be larger than the last finalized when they were created (we reject smaller ones), we don't control the order in which entries get finalized, and so discard these
 - given the above prune, it should be impossible for us to finalize an index that is smaller than the last one we saw (because we've pruned those entries)

benesjan replied:
Based on your previous comment I flattened it [here](https://github.com/AztecProtocol/aztec-packages/commit/a7123fb84a56fdd1d4f4aafe893715b6951e4989) which makes it a bit less twisted.

> currentFinalized is a confusing name for something we also call lastFinalized.

Renamed to lastFinalized

> given the above prune, it should be impossible for us to finalize an index that is smaller than the last one

Good point. Will throw there an error instead.

https://github.com/AztecProtocol/aztec-packages/pull/17611/commits/e4ff5c651a7691755e7af23ed102eabc57af9b16

nventuro commented on yarn-project/pxe/src/storage/tagging_data_provider/sender_tagging_data_provider.test.ts:
```diff
@@ -0,0 +1,444 @@
+      const txHash2 = TxHash.random();
+
+      // First, finalize an index
+      await taggingDataProvider.storePendingIndexes([{ secret: secret1, index: 3 }], txHash1);
+      await taggingDataProvider.updateStatusToFinalized(txHash1);
+
+      // Then add a higher pending index
+      await taggingDataProvider.storePendingIndexes([{ secret: secret1, index: 7 }], txHash2);
+
+      const lastUsed = await taggingDataProvider.getLastUsedIndex(secret1);
+      expect(lastUsed).toBe(7);
```
Had you finalized txHash2 instead of 1, this should still return 7 (but it will return 3). This is the bug I pointed out above.

benesjan replied:
I don't follow. When I finalize tx hash 2 it returns 7

The code always returns the highest pending or latest finalized

nventuro replied:
I believe it's because I did not realize the pending were being pruned.

nventuro commented on yarn-project/pxe/src/tagging/sync/sync_sender_tagging_indexes.ts:
```diff
@@ -0,0 +1,121 @@
+// tagging secret. If we get more tag indexes consumed than this window, an error is thrown in `PXE::proveTx` function.
+// This is set to a larger value than MAX_PRIVATE_LOGS_PER_TX (currently 32) because there could be more than
+// MAX_PRIVATE_LOGS_PER_TX indexes consumed in case the logs are squashed. This happens when the log contains a note
+// and the note is nullified in the same tx.
```
I don't understand what this comment is trying to say. Does it mean we cannot find a correct length?

benesjan replied:
It means I had to set it somewhat arbitrarily. I initially thought I can set it to MAX_PRIVATE_LOGS_PER_TX but it doesn't work because there can be more tags used than that due to squashing.

nventuro commented on yarn-project/pxe/src/tagging/sync/sync_sender_tagging_indexes.ts:
```diff
@@ -0,0 +1,121 @@
+// Note: Set it to 95 because `e2e_pending_note_hashes_contract` test hit 95 indexes emitted and 100 makes
+// `docs_examples.test.ts` fail as our JSON RPC doesn't seem to allow arrays with length larger than 99. I (benesjan)
+// think this window length should be fine.
+export const WINDOW_LEN = 95;
```
It's going to be problematic to change this value in the future, so we should have a clear understanding of why we chose it. What is `docs_examples.test.ts` and why does it hit 95 indices? Is that reasonable? Our RPC not handling more than 100 at a time is not good rationale. They could reduce page sizes and we'd need to deal with it.

benesjan replied:
Expanded on the choice in [this commit](https://github.com/AztecProtocol/aztec-packages/pull/17611/commits/216e84d1ec72aa47e994c370815c877ae1d96f59).

It would be problematic changing it but only in the down direction as that could then break people's apps. So I would say it makes sense to aim to have it be lower. But 95 is the minimum we can have without breaking the overflow tests (see the comment) so IDK.

I would say this value is fine because we can optimize the Aztec Node if it becomes a problem.

nventuro commented on yarn-project/pxe/src/tagging/sync/sync_sender_tagging_indexes.ts:
```diff
@@ -0,0 +1,121 @@
+export const WINDOW_LEN = 95;
+
+/**
+ * Syncs the highest finalized tagging index and pending tagging indexes for a given secret.
```
This is a bit too low level, I'd just say that it syncs the tagging state or whatever, and expand later what this means in detail. 

It makes this function look like an internal utility as it mentions lots of low level detail, which it very much is not. It's even missing important information such as _when_ we should call it.

benesjan replied:
https://github.com/AztecProtocol/aztec-packages/pull/17611/commits/1b47695bdc7597f1a01a2a20f8dc13cc4ffffc4f

nventuro commented on yarn-project/pxe/src/tagging/sync/sync_sender_tagging_indexes.ts:
```diff
@@ -0,0 +1,121 @@
+  // storing all those indexes as pending, and then checking the status of each pending index to update its
+  // finalization status accordingly.
+
+  const finalizedIndex = await taggingDataProvider.getLastFinalizedIndex(secret);
```
```suggestion
  const finalizedIndex = await taggingDataProvider.getLastFinalizedIndex(secret) ?? 0;
```

benesjan replied:
This is not correct because 2 lines below I do:

`let start = finalizedIndex === undefined ? 0 : finalizedIndex + 1;`

and there I need to know whether the index was undefined to start from 0 or defined to start from `last finalized + 1`

nventuro commented on yarn-project/pxe/src/tagging/sync/sync_sender_tagging_indexes.ts:
```diff
@@ -0,0 +1,121 @@
+
+    // Retrieve all indexes within the current window from storage. For each secret-txHash pair, we only store the
+    // highest index in the data provider. This means we may not get a transaction hash even if some of its indexes
+    // fall within the window (specifically when `end` is greater than those indexes). This is fine as those tx hashes
```
I don't understand this paragraph, what does 'its' refer to?

benesjan replied:
This was confusing. 'its` referred to indexes within the tx.

I decided to just drop most of the comment. That info was not really needed there

https://github.com/AztecProtocol/aztec-packages/pull/17611/commits/f1c5d80adf0136b09e8192643c07e9cc9029c5d7

nventuro commented on yarn-project/pxe/src/tagging/sync/sync_sender_tagging_indexes.ts:
```diff
@@ -0,0 +1,121 @@
+  let previousFinalizedIndex = finalizedIndex;
+  let newFinalizedIndex = undefined;
+
+  while (true) {
+    // Load and store indexes for the current window. These indexes may already exist in the database if txs using
+    // them were previously sent from this PXE. Any duplicates are handled by the tagging data provider.
```
This whole structure is quite complicated, could probably use with one or two helper functions. It's also quite untestable as-is.

benesjan replied:
Moved the logic related to status change to a separate func in https://github.com/AztecProtocol/aztec-packages/pull/17611/commits/db842b6f39708586082f606cc595d7d8b3ad2440

nventuro commented on yarn-project/pxe/src/tagging/sync/sync_sender_tagging_indexes.ts:
```diff
@@ -0,0 +1,121 @@
+  while (true) {
+    // Load and store indexes for the current window. These indexes may already exist in the database if txs using
+    // them were previously sent from this PXE. Any duplicates are handled by the tagging data provider.
+    await loadAndStoreNewTaggingIndexes(secret, app, start, end, aztecNode, taggingDataProvider);
```
A broad-strokes explanation of wtf is happening would be extremely useful. My tldr is something like

- we search for all used tags from the last tip into the future
- we check if any of the tags we found was finalized, which moves the tip forward
- repeat until the tip doesnt move forward
- record the information we found

This nicely showcases that moving the window forward requires 2 queries (logs + tx receipt), so if going from e.g. index 0 to 500 with a window length of 100, that'll take 10 roundrips (5*2) at least.

benesjan replied:
https://github.com/AztecProtocol/aztec-packages/pull/17611/commits/a55fa6bb56bfc34d4ca88e5dfdbcec7520a7ae6a

nventuro commented on yarn-project/pxe/src/pxe.ts:
```diff
@@ -735,14 +743,33 @@ export class PXE {
+            if (preTag.index > finalizedIndex + SENDER_TAGGING_INDEXES_SYNC_WINDOW_LEN) {
+              throw new Error(
+                `Highest used index ${preTag.index} is further than window length from the highest finalized index ${finalizedIndex}. Tagging window length ${SENDER_TAGGING_INDEXES_SYNC_WINDOW_LEN} is configured too low. Contact the Aztec team to increase it!`,
+              );
```
See how this makes the calsite unnecessarily complicated. It should just be 'hey bro here are the tags i used in this tx tyvm'

benesjan replied:
Good point. This is indeed incorrectly placed.

Moved the check to `storePendingIndexes` in  https://github.com/AztecProtocol/aztec-packages/pull/17611/commits/e07d79d3e665468085b2cfd94dd4f57cb81dc8a0

nventuro reviewed (commented):
Left a lot of comments since this is a very large PR. In hindsight we probably should've implemented some of the logic e.g. re. to pending or to processing in separate prs, and then merged that with pxe's work. So unfortutunately there's lots of comments both big and small in the mix.

Sorry it took so long to review this. Thanks for the effort in keeping it up to date and as understandable as possible.

nventuro reviewed (approved):
The improvements are excellent, thank you very much for going through my comments and considering each of them.

benesjan committed:
refactor: more robust tagging index sync as sender

In this PR I implement the approach that Nico originally described in [this post](https://forum.aztec.network/t/on-note-discovery-and-index-coordination/7165) for sender tagging index sync (for recipient it will be done in a followup PR).

To summarize I do the following:
1. Associate the tagging index with a hash of a tx in which the corresponding tax was used,
2. I track the status of the given tx in the tagging data provider,
3. tagging index status can be either PENDING, FINALIZED or DROPPED_OR_REVERTED (this status is only symbolic as when we realize a corresponding tx has been dropped or reverted we just delete the pending index),
4. when choosing the next index to use when sending a private log I choose `HIGHEST_PENDING_INDEX + 1` (or `HIGHEST_FINALIZED_INDEX + 1` if there is no pending index),
5. if the chosen index is further than WINDOW_LENGTH away from the HIGHEST_FINALIZED_INDEX an error is thrown. This should be a good guarantee of us never losing a log because we are always looking for a WINDOW_LENGTH of logs.

# Notes for reviewer
- I separated `TaggingDataProvider` into `SenderTaggingDataProvider` and `RecipientTaggingDataProvider` because the algorithms are completely disjoint and it makes the code clearer.
- Moved the `syncTaggedLogsAsSender` function to `pxe/src/tagging` directory and renamed it as `syncSenderTaggingIndexes`. Now `PrivateExecutionOracle` just calls this function. This is a step in the direction of deprecating `PXEOracleInterface` and it allowed me to write a reasonably clear unit tests.

# Issues created in this PR
- https://github.com/AztecProtocol/aztec-packages/issues/17775
- https://github.com/AztecProtocol/aztec-packages/issues/17776

Co-authored-by: Jan Beneš <janbenes1234@gmail.com>
